src/path_manager.c

```diff
@@ -721,6 +749,137 @@ static void handle_mptcp_event(struct l_genl_msg *msg, void *user_data)
+ *       kernel.
+ */
+static bool check_mptcp_socket_support(void)
+{
+#if !HAVE_DECL_IPPROTO_MPTCP
+# define IPPROTO_MPTCP 262
+#endif  // !HAVE_DECL_IPPROTO_MPTCP
+
+        int const fd = socket(AF_INET, SOCK_STREAM, IPPROTO_MPTCP);
+
+        // An errno other than EINVAL is unexpected.
+        if (fd != -1) {
+                close(fd);
+                return true;
+        } else if (errno != EINVAL) {
+                l_error("Unable to confirm MPTCP socket support.");
+        }
+
+        return false;
+}
+
+/**
+ * @brief Verify that MPTCP is enabled at run-time in the kernel.
+ *
+ * Check that MPTCP is enabled through the @c net.mptcp.mptcp_enabled
+ * @c sysctl variable if it exists.
+ *
+ * @note The @c net.mptcp.mptcp_enabled is supported by the
+ *       multipath-tcp.org kernel.
+ */
+static bool check_kernel_mptcp_enabled(void)
+{
+        static char const mptcp_enabled[] =
+                MPTCP_SYSCTL_BASE "mptcp_enabled";
+
+        FILE *const f = fopen(mptcp_enabled, "r");
+
+        if (f == NULL)
+                return false;  // Not using multipath-tcp.org kernel.
+
+        int enabled = 0;
+        int const n = fscanf(f, "%d", &enabled);
+
+        fclose(f);
+
+        if (likely(n == 1)) {
+                // "enabled" could be 0, 1, or 2.
+                enabled = (enabled == 1 || enabled == 2);
+
+                if (!enabled) {
+                        l_error("MPTCP is not enabled in the kernel.");
+                        l_error("Try 'sysctl -w "
+                                "net.mptcp.mptcp_enabled=2'.");
+
+                        /*
+                          Mptcpd should not set this variable since it
+                          would need to be run as root in order to
+                          gain write access to files in
+                          /proc/sys/net/mptcp/ owned by root.
+                          Ideally, mptcpd should not be run as the
+                          root user.
+                        */
+                }
+        } else {
+                l_error("Unable to determine if MPTCP is enabled.");
+        }
+
+        return enabled;
+}
+
+/**
+ * @brief Verify that the MPTCP "netlink" path manager is selected.
+ *
+ * Mptcpd requires MPTCP the "netlink" path manager to be selected at
+ * run-time when using the multipath-tcp.org kernel.  Issue a warning
+ * if it is not selected.
+ */
+static void check_kernel_mptcp_path_manager(void)
+{
+        static char const mptcp_path_manager[] =
+                MPTCP_SYSCTL_BASE "mptcp_path_manager";
+
+        FILE *const f = fopen(mptcp_path_manager, "r");
+
+        if (f == NULL)
+                return;  // Not using multipath-tcp.org kernel.
+
+        char pm[MPTCP_PM_NAME_MAX + 1] = { 0 };
+
+        static char const MPTCP_PM_NAME_FMT[] =
+                "%" L_STRINGIFY(MPTCP_PM_NAME_MAX) "s";
+
+        int const n = fscanf(f, MPTCP_PM_NAME_FMT, pm);
+
+        fclose(f);
+
+        if (likely(n == 1)) {
```

Comment:
The `likely` compiler hint seems unnecessary on this not-performance-sensitive code path. Not a big deal, though.

Reply:
Good point.  I dropped it commit a35bad9.